float32 object
float32 goal
